### PR TITLE
implement fill match arm assist for tuple of enums

### PR DESCRIPTION
**Pull request by JoshMcguigan**
This updates the fill match arm assist to work in cases where the user is matching on a tuple of enums. 

Note, for now this does not apply when some match arms exist (other than the trivial `_`), but I think this could be added in the future.

I think this also lays the groundwork for filling match arms when matching on tuples of non-enum values, for example a tuple of an enum and a boolean.

### CODE REVIEW

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 1]
```diff
@@ -54,13 +49,40 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
...
+            .filter(|variant_pat| is_variant_missing(&mut arms, variant_pat))
+            .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()))
+            .collect()
+    } else if let Some(enum_defs) = resolve_tuple_of_enum_def(&ctx.sema, &expr) {
+        // partial fill not currently supported for tuple of enums
+        if !arms.is_empty() {
+            return None;
+        }
+
+        enum_defs
```
Comment (edwin0cheng) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: Personally, since using `multi_cartesian_product` is quite convoluted, I think it will be nice to add a comment here describing what is going on here.

File: crates/ra_assists/Cargo.toml [hunk 2]
```diff
@@ -11,6 +11,7 @@ doctest = false
 format-buf = "1.0.0"
 join_to_string = "0.1.3"
 rustc-hash = "1.1.0"
+itertools = "0.8.2"
```
Comment (SomeoneToIgnore) [inline crates/ra_assists/Cargo.toml]: There's a `0.9.0` version out in the wild, maybe we can use that, just to be up to date here at least.

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 3]
```diff
@@ -54,13 +49,40 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
...
+            .into_iter()
+            .map(|enum_def| enum_def.variants(ctx.db))
+            .multi_cartesian_product()
+            .map(|variants| {
+                let patterns = variants
+                    .into_iter()
+                    .filter_map(|variant| build_pat(ctx.db, module, variant))
+                    .collect::<Vec<_>>();
```
Comment (SomeoneToIgnore) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: ` .collect::<Vec<_>>()`

Nit: You don't really need to collect anything here and can pass the iterator directly into the `make::tuple_pat`

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 4]
```diff
@@ -104,6 +126,22 @@ fn resolve_enum_def(sema: &Semantics<RootDatabase>, expr: &ast::Expr) -> Option<
...
+    sema: &Semantics<RootDatabase>,
+    expr: &ast::Expr,
+) -> Option<Vec<hir::Enum>> {
+    Some(
+        sema.type_of_expr(&expr)?
+            .tuple_fields(sema.db)
+            .iter()
+            .map(|ty| match ty.as_adt() {
+                Some(Adt::Enum(e)) => e,
+                _ => panic!("handle the case of tuple containing non-enum"),
```
Comment (SomeoneToIgnore) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: I don't think panicking during assist construction just because we did not implement something is a good idea: first of all, it pollutes the logs with big traces that do not say anything meaningful besides that the developers did not cover all cases (and this might be frequent, since this construction also happens on hover). 
Also, with this panic, calling the `Quick fix` shortcut will show endless progress bar in VS Code, which is a bit distracting :)

So I propose to simply match the only case here and maybe leave a comment about the rest of the cases.

File: crates/ra_syntax/src/ast/make.rs [hunk 5]
```diff
@@ -136,6 +136,15 @@ pub fn placeholder_pat() -> ast::PlaceholderPat {
     }
 }
 
+pub fn tuple_pat(pats: impl IntoIterator<Item = ast::Pat>) -> ast::TuplePat {
+    let pats_str = pats.into_iter().map(|p| p.syntax().to_string()).join(", ");
```
Comment (SomeoneToIgnore) [inline crates/ra_syntax/src/ast/make.rs]: Nit: `p.syntax().to_string()` can be just `p.to_string()` due to the recent improvements in the ast API.

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 6]
```diff
@@ -307,6 +345,103 @@ mod tests {
         );
     }
 
+    #[test]
```
Comment (SomeoneToIgnore) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: Currently, the assist panics on a single element tuple:

```rust
match (Some(22), ) {   
}
```

Might be a good idea to fix it :)

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 7]
```diff
@@ -104,6 +126,22 @@ fn resolve_enum_def(sema: &Semantics<RootDatabase>, expr: &ast::Expr) -> Option<
     })
 }
 
+fn resolve_tuple_of_enum_def(
```
Comment (SomeoneToIgnore) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: `resolve_enum_def` contains `autoderef` invocation.
This helps to detect more enum-related cases and might be a cool thing to support in `resolve_tuple_of_enum_def` too, now or later in some other PR.

Currently this code
```rust
let ref_opt = &Some(22);
match (ref_opt, ref_opt) {
}
```

will invoke the `panic!("handle the case of tuple containing non-enum"),` line and fail since we don't recognise the enum behind the reference.

This all kind of hints that we might want to approach the resolution somehow more unified, but that is maybe an another task outside of this PR's scope.

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 use std::iter;
 
+use itertools::Itertools;
+
```
Comment (matklad) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: We generally treat all non-std external crates the same, so there's no need for a blank line here.

File: crates/ra_assists/src/handlers/fill_match_arms.rs [hunk 9]
```diff
@@ -54,13 +49,40 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
...
+            .multi_cartesian_product()
+            .map(|variants| {
+                let patterns = variants
+                    .into_iter()
+                    .filter_map(|variant| build_pat(ctx.db, module, variant))
+                    .collect::<Vec<_>>();
+                ast::Pat::from(make::tuple_pat(patterns))
+            })
+            .filter(|variant_pat| is_variant_missing(&mut arms, variant_pat))
+            .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()))
```
Comment (Veetaha) [inline crates/ra_assists/src/handlers/fill_match_arms.rs]: This is not strictly related to this PR, but imho it would be better to put empty curly braces `{}` instead of a unit literal since the user almost all the time writes some block expression as a match arm (or a non-block expression which means removing the `(),` or `{}` anyway).

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 1 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: I added a comment here. Let me know if you prefer I say something different.

Comment (edwin0cheng) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Nice!

File: crates/ra_assists/Cargo.toml (hunk 2 above)
Comment (JoshMcguigan) [reply crates/ra_assists/Cargo.toml]: It seems all the other crates in this repo use `0.8.2`, so changing only this crate would mean compiling both versions. IMO it seems nice to use the same version of a dependency across all the crates in a single project like this, just to reduce mental overhead in cases where the API has actually changed. 

I'm not sure how dependency upgrades are typically done in this project, so let me know if there is some existing standard practice, but would you mind if I pushed this to a follow-up PR?

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 3 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Good catch, fixed.

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 4 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Oops, I mean to go back and fix that. Fixed.

File: crates/ra_syntax/src/ast/make.rs (hunk 5 above)
Comment (JoshMcguigan) [reply crates/ra_syntax/src/ast/make.rs]: Fixed.

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 6 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: This one was trickier than I thought. The panic comes from `make::tuple_pat`. I didn't dig much into it, but I think it is because a single element tuple doesn't get parsed as a `TuplePat`. 

For now I handle this case by marking the assist as non-applicable if `tuple_len < 2`. I hope that is acceptable for now.

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 7 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Thanks for pointing this out. I've added the `autoderef` to `resolve_tuple_of_enum_def`, but I agree it does feel like there should be some way to reduce some of the duplication here.

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 8 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Fixed.

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 9 above)
Comment (JoshMcguigan) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: I did see this suggestion in #3687, but I'd prefer to handle that in a separate PR if you don't mind?

Comment (Veetaha) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Sure!

File: crates/ra_assists/src/handlers/fill_match_arms.rs (hunk 4 above)
Comment (matklad) [reply crates/ra_assists/src/handlers/fill_match_arms.rs]: Note that you can use `todo!()` or a `//TODO` comment for such cases: CI would prevent mering code with todos into master.

**Comment by JoshMcguigan** [comment]
I believe I've cleaned up / fixed all review feedback. I'm not sure who typically presses the 'Resolve conversation` button.

Let me know if you'd like to see anything else before this merges.

**Comment by lnicola** [comment]
> I'm not sure who typically presses the 'Resolve conversation` button.

GitHub does that when you change the code in a newer commit.

**Comment by SomeoneToIgnore** [comment]
> GitHub does that when you change the code in a newer commit.

That only works for me when I apply the code suggestions.

> I'm not sure who typically presses the 'Resolve conversation` button.

Afaik, it's visible only for you, here's my version of the interface, with no button at all:
image

But there's a limited number of persons who can write `bors r+` and fire the final check & merge, that's what important.

**Comment by matklad** [comment]
bors r+

Thanks!

To reiterate, if you feel like this, moving this to hir (somewhere around [here](https://github.com/rust-analyzer/rust-analyzer/blob/b605271d7f3fa3fd3ac4dd0e1520b80b5fb13b40/crates/ra_hir_ty/src/diagnostics.rs)) as a proper exhaustive match checking, that would be awesome. This will probably require reading rustc source code to learn how to do this properly though.

**Comment by JoshMcguigan** [comment]
> bors r+
> 
> Thanks!
> 
> To reiterate, if you feel like this, moving this to hir (somewhere around [here](https://github.com/rust-analyzer/rust-analyzer/blob/b605271d7f3fa3fd3ac4dd0e1520b80b5fb13b40/crates/ra_hir_ty/src/diagnostics.rs)) as a proper exhaustive match checking, that would be awesome. This will probably require reading rustc source code to learn how to do this properly though.

Thanks for the merge. I'll take a look at moving this to hir after tackling some of the low hanging fruit follow ups mentioned in this PR.
